[silgen_cleanup] Add simplifications for destructures/tuple round trips and Rvalue tuple martialling

gottesmm opened the pull request:
This is a pattern that comes up often when emitting code for homogeneous tuples. This just eliminates the pattern really early after SILGen to clean up/simplify the IR. All 4 transforms are tied very closely to SILGen so we do a lot of checks to restrict it to the relevant patterns. I outline the four cases below. NOTE: I split the commits into two based off of whether I was pattern matching the tuple or a destructure just since it made it easier to put it into CanonicalizeInstruction.

```
Case 1:

  hat(%x) = destructure_tuple(%x)
  %x2 = tuple(hat(%x))
    ->
  %x2 = %x

Case 2:

  %x = tuple(hat(%x))
  hat(%x2) = destructure_tuple x
    ->
  %x2 = %x

Case 3:

  %alloc = alloc_stack $(Klass, Klass)
  ...
  %t0 = tuple_element_addr %alloc, 0
  %t1 = tuple_element_addr %alloc, 1
  (%0, %1) = destructure_tuple %x
  ...
  store %0 to %t0
  store %1 to %t1
    ->
  %alloc = alloc_stack $(Klass, Klass)
  ...
  store %x to %alloc

Case 4:

  %alloc = alloc_stack $(Klass, Klass)
  ...
  (%0, %1) = destructure_tuple %x
  ...
  %t0 = tuple_element_addr %alloc, 0
  store %0 to %t0
  %t1 = tuple_element_addr %alloc, 1
  store %1 to %t1
    ->
  %alloc = alloc_stack $(Klass, Klass)
  ...
  store %x to %alloc
```

gottesmm committed:
[canonicalize-inst] Add optimization to eliminate tuple(destructure(x))

We require the destructure to only be used by the singular tuple. This is a
pattern that shows up when we begin to emit homogeneous tuples.

gottesmm committed:
[canonicalize-inst] Add a few small optimizations to eliminate destructure traffic from exploding RValues.

We handle two transformations here that occur when SILGen emits tuple RValues.

The first is destructure(tuple()) elimination:

  %x = tuple(hat(%x))
  hat(%x2) = destructure_tuple x
  ->
  %x2 = %x

The second is eliminating destructures patterns to initialize tuple
memory. Specifically we transform the following two patterns:

   %alloc = alloc_stack $(Int, Int)
   ...
   %t0 = tuple_element_addr %alloc, 0
   %t1 = tuple_element_addr %alloc, 1
   (%0, %1) = destructure_tuple %x
   ...
   store %0 to %t0
   store %1 to %t1

   %alloc = alloc_stack $(Int, Int)
   ...
   (%0, %1) = destructure_tuple %x
   ...
   %t0 = tuple_element_addr %alloc, 0
   store %0 to %t0
   %t1 = tuple_element_addr %alloc, 1
   store %1 to %t1

To:

  %alloc = alloc_stack $(Int, Int)
  ...
  store %x to %alloc

gottesmm commented:
Failure was due to named tuple elements. I just put in some logic to skip that.

eeckstein commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -508,6 +511,319 @@ tryEliminateUnneededForwardingInst(SILInstruction *i,
+//                   DestructureTuple <-> Tuple Round Trip
+//===----------------------------------------------------------------------===//
+
+/// hat(%x) = destructure_tuple x
```
Can you write `(%x0,...,%xn)` instead of `hat()`? It's easier to understand.

eeckstein commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -508,6 +511,319 @@ tryEliminateUnneededForwardingInst(SILInstruction *i,
+/// %x2 = tuple(hat(%x))
+/// ->
+/// %x2 = %x
+static SILBasicBlock::iterator
```
Not really the point of this PR, but I suggest to do it now: Can you use the `InstructionDeleter` in `CanonicalizeInstruction` to get rid of this iterator updating/passing/returning madness?

eeckstein commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -508,6 +511,319 @@ tryEliminateUnneededForwardingInst(SILInstruction *i,
+//                        Destructure Simplifications
+//===----------------------------------------------------------------------===//
+
+// %x = tuple(hat(%x))
```
`hat` -> `(x0,...,xn)`

gottesmm replied:
I am fine with this.

eeckstein commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -508,6 +511,319 @@ tryEliminateUnneededForwardingInst(SILInstruction *i,
+    return nextII;
+
+  // Make sure our tuple only has a single use, us.
+  if (!tupleInst->hasOneUse())
```
Do we care about `debug_value` uses? If yes: `hasOneUse` -> `hasOneNonDebugUse`

eeckstein commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -508,6 +511,319 @@ tryEliminateUnneededForwardingInst(SILInstruction *i,
+canonicalizeAwayTupleEltAddrDestructurePairs(DestructureTupleInst *dti,
+                                             SILBasicBlock::iterator next,
+                                             CanonicalizeInstruction &pass) {
+  if (dti->getModule().getStage() != SILStage::Raw)
```
Is it really required to be in raw stage? It would probably not harm to try this transformation also on canonical SIL. On the other hand, it's really a very specific pattern matching, which is unlikely to be generated by an optimization - I guess.
Anyway, can you add a comment why this is done only on raw SIL?

gottesmm replied:
The reason why I did this only in Raw SIL is exactly due to what you said. I can add a comment explaining that it is b/c it is so specific to the SILGen output that it probably won't be useful otherwise. I was just trying to save compile time by early exiting.

eeckstein commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -508,6 +511,319 @@ tryEliminateUnneededForwardingInst(SILInstruction *i,
+
+  auto results = dti->getResults();
+  SILValue firstResult = results[0];
+  auto *firstResultUse = firstResult->getSingleUse();
```
`getSingleNonDebugUser`?

eeckstein commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -508,6 +511,319 @@ tryEliminateUnneededForwardingInst(SILInstruction *i,
+    firstInst = si;
+    storeOwnershipQualifier = si->getOwnershipQualifier();
+    firstTEAI = dyn_cast<TupleElementAddrInst>(si->getDest());
+  }
```
```
if () {
} else if () {
} else {
  return
}
```
would be simpler, IMO

eeckstein commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -508,6 +511,319 @@ tryEliminateUnneededForwardingInst(SILInstruction *i,
+  for (auto ii = std::next(firstInst->getIterator(), 1),
+            ie = firstInst->getParent()->end();
+       ii != ie && numTupleEltStoreSeen < maxNumElements; ++ii) {
+    // We go down this path if we are in a tuple_element_addr, store,
```
That comment should be inside the `if`, I guess

eeckstein commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -517,6 +833,22 @@ CanonicalizeInstruction::canonicalize(SILInstruction *inst) {
 
+  if (auto *tuple = dyn_cast<TupleInst>(inst)) {
+    if (auto tupleTy = tuple->getType().getAs<TupleType>()) {
+      if (!tupleTy->hasAnyNamedElement()) {
```
Out of curiosity: what's the problem with named tuple elements?

gottesmm replied:
@eeckstein as an example, consider the following SIL:

```
// thunk for @callee_guaranteed (@unowned KeyPathBuffer) -> (@unowned UnsafeMutablePointer<B>, @owned Swift.AnyObject?, @error @owned Error)
sil shared [transparent] [serializable] [reabstraction_thunk] [ossa] @$ss13KeyPathBufferVSpyq_GyXlSgs5Error_pIgydozo_AbC7pointer_AD5ownertsAE_pIegyrzo_r0_lTR : $@convention(thin) <Root, Value> (KeyPathBuffer, @noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error)) -> (@out (pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), @error Error) {
// %0                                             // users: %4, %3
// %1                                             // user: %5
// %2                                             // user: %5
bb0(%0 : $*(pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), %1 : $KeyPathBuffer, %2 : $@noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error)):
  %3 = tuple_element_addr %0 : $*(pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), 0 // user: %9
  %4 = tuple_element_addr %0 : $*(pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), 1 // user: %10
  try_apply %2(%1) : $@noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error), normal bb1, error bb2 // id: %5

// %6                                             // user: %7
bb1(%6 : @owned $(UnsafeMutablePointer<Value>, Optional<AnyObject>)): // Preds: bb0
  (%7, %8) = destructure_tuple %6 : $(UnsafeMutablePointer<Value>, Optional<AnyObject>) // users: %9, %10
  store %7 to [trivial] %3 : $*UnsafeMutablePointer<Value> // id: %9
  store %8 to [init] %4 : $*Optional<AnyObject>   // id: %10
  %11 = tuple ()                                  // user: %12
  return %11 : $()                                // id: %12

// %13                                            // user: %14
bb2(%13 : @owned $Error):                         // Preds: bb0
  throw %13 : $Error                              // id: %14
} // end sil function '$ss13KeyPathBufferVSpyq_GyXlSgs5Error_pIgydozo_AbC7pointer_AD5ownertsAE_pIegyrzo_r0_lTR'
```

This we (before some changes I am going to make real quick) we would transform to:

```
sil shared [transparent] [serializable] [reabstraction_thunk] [ossa] @$ss13KeyPathBufferVSpyq_GyXlSgs5Error_pIgydozo_AbC7pointer_AD5ownertsAE_pIegyrzo_r0_lTR : $@convention(thin) <Root, Value> (KeyPathBuffer, @noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error)) -> (@out (pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), @error Error) {
// %0                                             // user: %5
// %1                                             // user: %3
// %2                                             // user: %3
bb0(%0 : $*(pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), %1 : $KeyPathBuffer, %2 : $@noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error)):
  try_apply %2(%1) : $@noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error), normal bb1, error bb2 // id: %3

// %4                                             // user: %5
bb1(%4 : @owned $(UnsafeMutablePointer<Value>, Optional<AnyObject>)): // Preds: bb0
  store %4 to [init] %0 : $*(pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>) // id: %5
  %6 = tuple ()                                   // user: %7
  return %6 : $()                                 // id: %7

// %8                                             // user: %9
bb2(%8 : @owned $Error):                          // Preds: bb0
  throw %8 : $Error                               // id: %9
} // end sil function '$ss13KeyPathBufferVSpyq_GyXlSgs5Error_pIgydozo_AbC7pointer_AD5ownertsAE_pIegyrzo_r0_lTR'
```

The verifier throws since the types do not match up. We really shouldn't care about tuple element names at the SIL level and should in type lowering eliminate them IMO. But that is more than I want to do in this PR (which is meant to only solve problems exposed by homogeneous tuples).

gottesmm replied:
You are right. I found a way to do it by checking if the types match. Thanks!

eeckstein reviewed (commented):
basically lgtm, just a few comments

BTW, don't forget to swift-ci benchmark

gottesmm committed:
Add some additional comments as per review feedback.

gottesmm commented:
@swift-ci benchmark

gottesmm committed:
[canonical-inst] Do not eliminate destructures that are results of apply.

There are a bunch of passes that assume that applies with tuple results will
always have a destructure. In a certain sense the destructure is treated as part
of the apply itself.

We really should eliminate this behavior by making apply take multiple
results. Until we begin that transition though, rather than breaking a lot of
code and the SIL I am going to respect that.

gottesmm commented:
@swift-ci test

gottesmm commented:
@mikeash I am trying to reproduce the weak stress test failure on Linux 16.04 and I am not doing so. Do you have any thoughts on why this is happening? (Maybe I need to try to ssh into one of the bots). Is the test flaky in some way? Specifically the issue is:

PASS: Swift(linux-x86_64) :: Runtime/weak-reference-racetests.swift (2 of 2)

atrick commented:
I reviewed the review comments and feel like pointing out that bootstrapping/incomplete optimizations are fine as long as they don't break SIL invariants that we want to rely on elsewhere. Namely

- debug information can't affect -O output

- In OSSA, all multi-value calls should be used by a single destructure. That destructure should not be removed at least until OSSA lowering. It would be nice if this were part of the SIL verifier. With true multi-value calls, the destructure just goes away and everything else remains the same. We can check a `DestructureInstruction::expandsCallResults` API wherever the optimizers handles destructures. Introducing a separate opcode would probably interfere with analysis in several places.
